torch/autograd/functional.py

```diff
@@ -0,0 +1,712 @@
+
+def hessian(func, inputs, create_graph=False, strict=False):
+    r"""Function that computes the Hessian of a given scalar function.
+
+    Args:
+        func (function): a Python function that takes Tensor inputs and returns
+            a Tensor with a single element.
+        inputs (tuple of Tensor or Tensor): inputs to the function ``func``.
+        create_graph (bool, optional): If ``True``, the Hessian will be computed in
+            a differentiable manner. Note that when strict is ``False``, the result can not
+            require gradients or be disconnected from the inputs.
+            Defaults to ``False``.
+        strict (bool, optional): If ``True``, an error will be raised when we detect that there exists an input
+            such that all the outputs are independent of it. If ``False``, we return zeros as the vjp for
+            said inputs, which is the expected mathematical value.
+            Defaults to ``False``.
+
+    Returns:
+        Hessian (Tensor or a tuple of tuple of Tensor) if there are a single input,
+            this will be a single Tensor containing the Hessian for the input.
+            If it is a tuple, then the Hessian will be a tuple of tuples where
+            ``Hessian[i][j]`` will contain the Hessian of the ``i``th input
+            and ``j``th input with size the sum of the size of the ``i``th input plus
+            the size of the ``j``th input.
+
+    Example::
+
+        >>> def pow_reducer(x):
+        ...   return x.pow(3).sum()
+        >>> inputs = torch.rand(2, 2)
+        >>> hessian(pow_reducer, inputs)
+        tensor([[[[5.2265, 0.0000],
+                  [0.0000, 0.0000]],
+
+                 [[0.0000, 4.8221],
+                  [0.0000, 0.0000]]],
+
+
+                [[[0.0000, 0.0000],
+                  [1.9456, 0.0000]],
+
+                 [[0.0000, 0.0000],
+                  [0.0000, 3.2550]]]])
+
+        >>> hessian(pow_reducer, inputs, create_graph=True)
+        tensor([[[[5.2265, 0.0000],
+                  [0.0000, 0.0000]],
+
+                 [[0.0000, 4.8221],
+                  [0.0000, 0.0000]]],
+
+
+                [[[0.0000, 0.0000],
+                  [1.9456, 0.0000]],
+
+                 [[0.0000, 0.0000],
+                  [0.0000, 3.2550]]]], grad_fn=<ViewBackward>)
+
+
+        >>> def pow_adder_reducer(x, y):
+        ...   return (2 * x.pow(2) + 3 * y.pow(2)).sum()
+        >>> inputs = (torch.rand(2), torch.rand(2))
+        >>> hessian(pow_adder_reducer, inputs)
+        ((tensor([[4., 0.],
+                  [0., 4.]]),
+          tensor([[0., 0.],
+                  [0., 0.]])),
+         (tensor([[0., 0.],
+                  [0., 0.]]),
+          tensor([[6., 0.],
+                  [0., 6.]])))
+    """
+
+    tuple_inputs, inputs = _as_tuple(inputs, "inputs", "hessian")
+
+    def ensure_single_output_function(*inp):
+        out = func(*inp)
+        tuple_out, t_out = _as_tuple(out, "outputs of the user-provided function", "hessian")
```

Comment:
nit: We don't really need this check right? We just need to check if `torch.is_tensor(out)`

Reply:
The reason for this is:
- We want t_out to be able to use _check_requires_grad
- That gives nice error messages for non-Tensor returns

Reply:
Oh I see, _as_tuple gives a nice error message here. Thanks for the clarification, I agree with keeping it.